Various performance improvements

moffatman opened the pull request:
Should be twice as fast according to my use cases. Happy to discuss what you need for testing, (micro)benchmarking, preserving public API for old slow functions, or whatever else to merge it. I know I still need to do stuff like bump version, changelog, etc.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

moffatman committed:
Avoid repeated memory allocation

moffatman committed:
Only calculate fileInfo if generateSpans

moffatman committed:
Use a trie to match entities

moffatman committed:
Optimize _invalidUnicode hot path

moffatman committed:
Use charCodes instead of one-char Strings

moffatman committed:
Optimize toAsciiLowerCase() hot path

moffatman committed:
Fix unget assertion for surrogate pairs

moffatman committed:
Fix trie traversal

moffatman committed:
Remove span stuff from !generateSpans tests

moffatman committed:
Run the formatter

moffatman committed:
Put entities trie in compile-time const

It avoids runtime tree building. And lets the tree be shared between isolates (multiple MB)

kevmoo commented:
please run the formatter!

moffatman committed:
Format/analyze

moffatman commented:
Idk whether the new `trie.dart` is the right way... maybe something to wait for dart macros

kevmoo reviewed (changes requested):
need to run the formatter, it seems

moffatman committed:
run dart format

moffatman committed:
Make Charcode abstract/final

moffatman commented on lib/src/constants.dart:
```diff
@@ -393,10 +393,54 @@ const mathmlTextIntegrationPointElements = [
 ];
 
-const spaceCharacters = ' \n\r\t\u000C';
+class Charcode {
```
Thx done

moffatman committed:
Some future proofing

- Test that the [entities] and [entitiesTrieRoot] match
- Script to regenerate trie.dart if [entities] ever changes

mosuem commented:
Closing as the [dart-lang/html](https://github.com/dart-lang/html) repository is merged into the [dart-lang/tools](https://github.com/dart-lang/tools) monorepo. Please re-open this PR there!
